lib/penelope/nlp/intent_classifier.ex

```diff
@@ -41,9 +41,10 @@ defmodule Penelope.NLP.IntentClassifier do
   alias Penelope.ML.Pipeline
 
   @type model :: %{
-    tokenizer: [{atom, any}],
-    classifier: [{atom, any}],
-    recognizer: [{atom, any}]
+    tokenizer:   [{atom, any}],
```

Comment:
This is a bit confusing at first glance—the types of the model components are function return values?

Reply:
They are pipelines, which are a list of atom/model pairs executed in order. I have been meaning to create a pipeline typespec and reuse it here.

Reply:
Ah, got it; that answers my next question, then (why the detokenizer is a reversed tokenizer). The distinction between component and pipeline was what was throwing me; I think I might opt for some kind of convention-driven difference between the two, e.g. making the pipeline names verbs instead of nouns or adding a `_pipeline` suffix. I guess this matches other ML platforms' convention more closely, though, so carry on.

Reply:
Yeah, I didn't have to reverse it before because I only had one stage. And tokenization doesn't compose well, so I basically just threw in the digit splitter, but it should be fine.

Reply:
Yeah; I noticed that. Don't see a problem with it, though, since it's designed as a replacement for the other one. You _could_ compose tokenizers by tokenizing each token and flattening the list, or take the Lucene approach by working in terms of token "streams", but meh.

Reply:
I went down that path a bit - the problem is that composing in the detokenizer direction is hard, because you would have to be able to re-split the list at each stage, and I think that would introduce some loss.

One way to do it would be to decouple the replacers (punctuation, digits, etc.) from the final splitter.

Reply:
Or to tag the tokens with the tokenizer that produced them and building the detokenized string by only giving the detokenizers access to the list of tokens they or a subsequent tokenizer produced.

Reply:
Makes sense.